### PR TITLE
BUG: Do not skip value-based promotion path for large Python integers

### PR DESCRIPTION
**Pull request by seberg**
When we have:
1. Only Python scalars (int, float, complex) as inputs
2. One of the integers is too large for the default integer dtype

The "new" promotion path (which existed for a long time internally!)
would not produce the same result as the legacy value-based path.
However, my logic failed to check for the second part, so it did not
run the legacy path in this odd little case.

This fixes the issue reported in:
https://github.com/numpy/numpy/issues/13754#issuecomment-1155126639

### CODE REVIEW

**Comment by seberg** [comment]
Unfortunately, this has been around since NumPy 1.21, I think.  But calling `np.result_type` on *only* Python scalars (and then also with a very large integer) is probably strange enough that nobody really runs into it.

**Comment by charris** [comment]
Thanks Sebastian.

**Comment by mayalinetsky-kryon** [comment]
Hi guys, thanks for the help and sorry for not replying (I'm in a different time zone).
In which version is the fix implemented?

**Comment by seberg** [comment]
It will definitely be included in 1.23.0.

However, we are considering cleaning up the behaviors where NumPy inspects value to guess the result dtype.  (Unfortunately, they can lead to "magical" changes in results and cause headaches in some places.)
 
This particular case with large integers is special and it may well be that the code would not change in the end, or that at least: `np.result_type(np.array(2**32), np.array(2))` and `np.array([2**32, 2]).dtype` would not change.  (Either of those may be good workarounds for the issue on older NumPy versions.)

But, because of those plans, I am a very curious to hear about your use-case!  I am wondering what NumPy could offer you if the current version would _stop_ working.  For example, would a function (or similar pattern):
```
np.get_min_int_dtype([2, 2**23], allow_unsigned=False, allow_object=True)
``` 
serve that need?  I.e. a function that tells you which NumPy integer dtype is "sufficient" for a certain set of integer values.
That could even be spelled as `np.array([2, 2**32], dtype=MinIntegerOrObject)` or so...

**Comment by mayalinetsky-kryon** [comment]
@seberg I don't really have a use case for calling `np.result_type` with only python integers, I've discovered this bug accidentally while working with a package called sparse.
Originally, I thought `np.result_type` would give me the result dtype of the product of the input, which would help solve the overflow problem in the sparse package I've reported [here](https://github.com/pydata/sparse/issues/552).

Bottom line, even if this bug will stay in older versions, I'm not gonna be affected.
